Stabilize some banking stage tests

ryoqun opened the pull request:
#### Problem

Various banking stage tests are fragile.

#### Summary of Changes

Fixed several race conditions in tests.

This fixes 3 unstable unit tests. For all of them, the root cause of unstability is a race condition between solana-poh-service-tick_producer thread and solana-banking-stage-tx thread. So, to precisely control the timings of solana-poh-service-tick_producer thread, I introduced new PohConfig field just for testing purpose.

The requested explanations for each tests follow (in simpler-hardest order):

- `test_banking_stage_tick`
  - Problem: Even if we waited for 600ms, it's possible for ticks not to be produced enough when the ticker thread is stalled under heavy system load.
  - Solution: Always ensure enough ticks are produced (in this pull request, produce 6 ticks)
- `test_banking_stage_entryfication` (originally mentioned by the issue)
  - Problem: Right in the middle of solana-banking-stage-tx's processing of received packets, the thread may be paused and the solana-poh-service-tick_producer thread starts kicking in. If that is the case and new tick entry unfortunately knocks the maximum tick height, it interrupts the in-flight processing of the packets. Depending on the exact timing, the banking stage interrupt results in MaxHeightReached error from PohRecorder when banking stage tries to record a transaction entry or simply enqueuing code path of (unprocessed_packets=>buffered_packets). Then, at test code, this results in an infinite loop at HERE (btw, this loop is itself a bit off..., should fix it too?), and eventual timeout on the CI, vomitting an odd RocksDB error.
  - Solution: Prevent slots from becoming full merely with tick PoH entries.
- `test_banking_stage_entries_only`
  - Problem: In addition to the problem and solution of `test_banking_stage_entryfication`. There are additional issues. First, assertion may be run before banking stage finishes. Second, the wall time resulted from `for _ in 0..10` + `sleep(Duration::from_millis(200));` may not be enough. 
  - For the two additional issue, do simple fix: reorder `banking_stage.join().unwrap();` up in the test code and replace the 10-time repetition with unbounded looping also with guard for empty entries from `entry_receiver`

(First starting from #5660, I ended up with spotting other similar race conditions in neighbor tests, so fixed them as well.)

Fixes #5660

ryoqun commented on core/src/banking_stage.rs:
```diff
@@ -1128,18 +1140,20 @@ mod tests {
                     .map(|(_bank, (entry, _tick_height))| entry)
                     .collect();
 
                 assert!(entries.verify(&blockhash));
```
this assertion should be moved inside the newly-added empty guard if clause.

ryoqun commented on core/src/banking_stage.rs:
```diff
@@ -1128,18 +1140,20 @@ mod tests {
-                    bank.process_transactions(&entry.transactions)
-                        .iter()
-                        .for_each(|x| assert_eq!(*x, Ok(())));
+                if !entries.is_empty() {
```
Under high load, I observed empty entries are returned.

ryoqun commented on core/src/poh_service.rs:
```diff
@@ -32,7 +32,11 @@ impl PohService {
+                    if poh_config.target_tick_count.is_none() {
+                        Self::sleepy_tick_producer(poh_recorder, &poh_config, &poh_exit_);
+                    } else {
+                        Self::short_lived_tick_producer(poh_recorder, &poh_config, &poh_exit_);
```
naming: or `short_lived_sleepy_tick_producer`.. (too long?)

carllin replied:
I like descriptive names 👍

ryoqun commented on core/src/banking_stage.rs:
```diff
@@ -928,14 +928,19 @@ impl Service for BankingStage {
 ) {
     let exit = Arc::new(AtomicBool::new(false));
-    let poh_config = Arc::new(PohConfig::default());
+    let poh_config = if let Some(poh_config) = poh_config {
```
Hmm, I'm sure there is better way to write in more idiomatic Rust way.

carllin replied:
Maybe `let poh_config = Arc::new(poh_config.unwrap_or(PohConfig::default()))` :D

carllin commented:
@ryoqun, thanks for taking this on! In the `Problem` description of this PR, could you describe in greater detail the race condition that is making the test fragile? It would better frame the discussion for this change.

ryoqun commented:
@carllin Thanks for checking this out! Sure thing, I'll elaborate it! I'll work on to finish this up tomorrow's work-time in JST. :)

ryoqun commented:
@carllin Sorry for being late... I added the details of race conditions. Hope that clarifies what I wanted to fix... Also updated this PR with your comments!

carllin commented on core/src/banking_stage.rs:
```diff
@@ -1020,8 +1021,10 @@ mod tests {
                 Blocktree::open(&ledger_path).expect("Expected to be able to open database ledger"),
             );
+            let mut poh_config = PohConfig::default();
+            poh_config.target_tick_count = Some(6);
```
For clarity, instead of the magic number 6, could you write this as `genesis_block.ticks_per_slot + num_extra_ticks` for some number `num_extra_ticks`.

ryoqun replied:
sure thing! done at https://github.com/solana-labs/solana/pull/6251/commits/a9138d1efdcaff8d1b8979c638c13aefab0ed444

carllin replied:
You were right 😄 for the same reasons as the other cases listed below, this should also use `bank.max_tick_height()` instead of `genesis_block.ticks_per_slot` for clarity and consistency

ryoqun replied:
Nice catch! done cef4770bee4c27a70fe4001ae35dae29f148e984

carllin commented on core/src/banking_stage.rs:
```diff
@@ -1212,8 +1220,11 @@ mod tests {
                 );
+                let mut poh_config = PohConfig::default();
+                // limit the tick to 1 to prevent clearing working_bank at PohRecord then PohRecorderError(MaxHeightReached) at BankingStage
+                poh_config.target_tick_count = Some(1);
```
Good catch, quite a subtle bug, especially for your first time through the code base :)

Instead of the magic number 1, can we use something like `genesis_block.ticks_per_slot - 1`, to make it clear that there's nothing special about this number, it just needs to be < genesis_block.ticks_per_slot

ryoqun replied:
sure thing! done at https://github.com/solana-labs/solana/pull/6251/commits/a9138d1efdcaff8d1b8979c638c13aefab0ed444

ryoqun replied:
Well, I couldn't directly use genesis_block.ticks_per_slot because subtracting -1 from it still causes the race condition, so I chose different one: b07bf58e0cfc580f65e08ee731ce88fc2af8669e

carllin replied:
@ryoqun, ah yes, it's because of a special case where slot 0 generates one less tick than all the other slots. For instance, if `genesis_block.ticks_per_slot` is equal to 10, then the first slot will have `bank.max_tick_height` of `9`, so PoH will, starting at 0, generate a tick for tick_heights `0->1, 1->2,  2->3...8->9`, which is only 9 ticks. The next slot, slot 1, will then generate 10 ticks because it's max tick height is 19, and it will start generating a tick for tick heights `9->10, 10->11, ... 18 -> 19`. This is all because in the past we wanted to keep the first tick in a slot as 0-indexed. 

This open PR here actually will address this issue: https://github.com/solana-labs/solana/pull/6263.

ryoqun replied:
Thanks for very detailed explanation! I'll look at it. :D

carllin commented on core/src/banking_stage.rs:
```diff
@@ -1069,8 +1071,11 @@ mod tests {
             );
+            let mut poh_config = PohConfig::default();
+            // limit the tick to 1 to prevent clearing working_bank at PohRecord then PohRecorderError(MaxHeightReached) at BankingStage
+            poh_config.target_tick_count = Some(1);
```
Instead of the magic number 1, can we use something like genesis_block.ticks_per_slot - 1, to make it clear that there's nothing special about this number, it just needs to be < genesis_block.ticks_per_slot

ryoqun replied:
sure thing! done at https://github.com/solana-labs/solana/pull/6251/commits/a9138d1efdcaff8d1b8979c638c13aefab0ed444

ryoqun replied:
Well, I couldn't directly use genesis_block.ticks_per_slot because subtracting -1 from it still causes the race condition, so I chose different one: b07bf58e0cfc580f65e08ee731ce88fc2af8669e

carllin commented:
@ryoqun, looking good! Just a few nits, and a little more plumbing to make CI pass :)

ryoqun commented on core/src/banking_stage.rs:
```diff
@@ -1153,13 +1164,11 @@ mod tests {
     }
 
     #[test]
-    #[ignore]
```
Yay!
